services/src/main/java/org/keycloak/authentication/forms/RegistrationProfile.java

```diff
@@ -67,33 +64,37 @@ public void validate(org.keycloak.authentication.ValidationContext context) {
 
         context.getEvent().detail(Details.REGISTER_METHOD, "form");
 
-        UserProfileValidationResult result = forRegistrationProfile(context.getSession(), formData).validate();
-        List<FormMessage> errors = Validation.getFormErrorsFromValidation(result);
+        UserProfileProvider profileProvider = context.getSession().getProvider(UserProfileProvider.class);
+        UserProfile profile = profileProvider.forValidation(UserProfileContext.REGISTRATION_PROFILE, formData);
+
+        try {
+            profile.validate();
+        } catch (ValidationException pve) {
+            List<FormMessage> errors = Validation.getFormErrorsFromValidation(pve.getErrors());
 
-        if (errors.size() > 0) {
-            if (result.hasFailureOfErrorType(Messages.EMAIL_EXISTS, Messages.INVALID_EMAIL)) {
-                UserProfile updatedProfile = result.getProfile();
-                context.getEvent().detail(Details.EMAIL, updatedProfile.getAttributes().getFirstAttribute(UserModel.EMAIL));
+            if (pve.hasError(Messages.EMAIL_EXISTS, Messages.INVALID_EMAIL)) {
+                context.getEvent().detail(Details.EMAIL, profile.getAttributes().getFirstValue(UserModel.EMAIL));
             }
 
-            if (result.hasFailureOfErrorType(Messages.EMAIL_EXISTS)) {
+            if (pve.hasError(Messages.EMAIL_EXISTS)) {
                 context.error(Errors.EMAIL_IN_USE);
                 formData.remove("email");
```

Comment:
Why is email removed from the form if it is used? It is really not a good UX, common is to keep value in the form when it is reshown with the error message and let user to patch the value. But probably we need to create jira issue for this and patch it in special PR later

Reply:
This is not a new code. I did not touch anything that was already there.